### ISSUE TITLE
Bounce feature broken?

### ISSUE THREAD

**Issue by ryanfeeley** [mozilla/fxa-content-server#5629]
STEPS TO REPRODUCE
1. Register an account with an invalid mozilla email that will bounce (I went direct to accounts.firefox.com)

EXPECTED RESULTS
* Email bounces, and user is bounced back to registration form with Mistyped email warning

ACTUAL RESULTS
* Nothing happens

![image](https://user-images.githubusercontent.com/68213/31838401-a9039ace-b5aa-11e7-8859-d02bdcab3e8e.png)

**Comment by shane-tomlinson** [comment]
@shane-tomlinson to investigate

**Comment by vladikoff** [comment]
@jrgm what do we need to make this work? Did we switch something in our bouncing logic?

**Comment by vladikoff** [comment]
> switch something 

moved from `postfix`?

**Comment by vladikoff** [comment]
from mtg: @jrgm to check again in prod

**Comment by jrgm** [comment]
I sent an email to `alk33jfkjlkfjalkjflasjdflaksjflksjdfklj2@mozilla.com`.

It bounced. A bounce handler picked it up and logged:
```
"Fields":{
  "op":"handleBounce",
  "action":"failed",
  "email":"alk33jfkjlkfjalkjflasjdflaksjflksjdfklj2@mozilla.com",
  "domain":"other",
  "bounce":true,
  "status":"5.0.0",
  "template":"verifyEmail",
  "lang":"en",
  "bounceType":"Transient",
  "bounceSubType":"General"
}
```

But this section with `shouldDelete` would evaluate to false, no? No deletion attempted?
https://github.com/mozilla/fxa-auth-server/blob/master/lib/email/bounces.js#L137-L152

The `fxa.accounts` table still has an entry for `alk33jfkjlkfjalkjflasjdflaksjflksjdfklj2@mozilla.com`, that is unverified.

**Comment by vladikoff** [comment]
@rfk @jrgm  I don't think we can rely on `"bounceType":"Transient"` , or maybe if it is `"template":"verifyEmail",` and `transient` then we can still delete?

**Comment by jrgm** [comment]
Per https://docs.aws.amazon.com/ses/latest/DeveloperGuide/notification-contents.html#bounce-types, I'm not clear why this email Bounce object came back as `Temporary/General`. `Permanent/NoEmail` seems more appropriate. (Of course, SES can only tell what the remote MTA told them. So is this just corporate gmail?

Remember. The `S` in `SMTP` stands for `Simple`!

**Comment by rfk** [comment]
>  I don't think we can rely on "bounceType":"Transient"

Yeah, we deliberately don't delete on transient bounces, on the assumption that the user will fix whatever problem was happening and then trigger an email resend.

The whole "detecting a bounce by noticing that the account got deleted" thing is a bit of a hack.  If we want to do a more reliable job of this, we should probably do the work to introduce a first-class "your verification email bounced" state on the account.  But that's a big refactor :-/

**Comment by vladikoff** [comment]
> Yeah, we deliberately don't delete on transient bounces, on the assumption that the user will fix whatever problem was happening and then trigger an email resend.

Yeah that's why I was suggesting using it only on account creation. We would tell users to go fix the mailbox and re-register.

**Comment by rfk** [comment]
One possible risk is that AWS will retry delivery of a message that failed with a `Transient` bounce, so if we delete the account but the the delivery gets retried and succeeds, that could be a confusing user experience.  Then again, the user may be in a confused state already...

**Comment by rfk** [comment]
> is that AWS will retry delivery of a message that failed with a Transient bounce

The docs say "Amazon SES notifies you of hard bounces and soft bounces that will no longer be retried" so it sounds like this is not a risk in practice.

**Comment by jrgm** [comment]
Okay. I sent an fxa signup to get a verifySyncEmail email from: fastmail.com, gmail.com, hotmail.com, icloud.com, mail.com, mozilla.com, outlook.com, yahoo.com, yandex.com.

All of those domains, **except** mozilla.com, immediately resulted in a bounce notification with `"bounceType":"Permanent","bounceSubType":"General"` logged and the account was deleted.

For mozilla.com, first, a deliver notification was received and handled, followed one second later by a bounce notification with `"bounceType":"Transient","bounceSubType":"General"` logged. The account was **not** deleted.

**Comment by jrgm** [comment]
I sent email directly from me to a non-existent mailbox at @gmail.com and @mozilla.com

Both immediately returned bounce notifications.

Looking in detail at the bounce email, and locating the `Status:` header in the content-type message/delivery-status submessage the responses subtly differ.

Nosuchmailbox@mozilla.com gets `Status: 5.0.0` which is `Permanent - Other or Undefined Status` from rfc 3463 https://tools.ietf.org/html/rfc3463.

Nosuchmailbox@gmail.com gets `Status: 5.1.1` which is `Permanent - Bad destination mailbox address`.

So whose wrong here? Googlemail says the mozilla.com bounce is Permanent, but doesn't explain more. Does SES have some valid reason for re-interpreting the `Permanent` class designation?

**Comment by jrgm** [comment]
(Note: `Nosuchmailbox @ gmail.com` was not actually the email used.)

**Comment by jrgm** [comment]
And, maybe both are wrong ;-)

**Comment by jrgm** [comment]
For those of you that are curious, here are the various `Status` responses by domain:

Mail domain     Status
@gmail.com 	5.1.1
@fastmail.com 	5.1.1
@outlook.com 	5.5.0
@mail.com 	5.3.0
@mozilla.com 	5.0.0
@icloud.com 	5.1.1
@hotmail.com 	5.5.0
@yahoo.com 	5.3.0
@yandex.com 	5.7.1

(I dunno. Kinda hard to understand why someone would choose e.g., `Other or undefined protocol status` when `Bad destination mailbox address` is what happened).

**Comment by jrgm** [comment]
We *could* act on the `status` code if it has a leading `5.`. However, I did review of currently available bounce logs, and I saw a couple of examples of a `5.x.x` which SES classified as a `Transient` bounceType and probably should have done have classified it that way (based on my reading other details in the `diagnoticCode` field, e.g., authentication required, denied for policy reasons).

In other words, it's a messy world and SES probably has a pile of complex heuristics to (hopefully) interpret that mess.

**Comment by jrgm** [comment]
So to summarize, the bounce feature is **not** broken. When SES tells us that a bounce is permanent, we **do** delete the account (and the frontend responds correctly when it detects the account is gone).

If a remote MTA does not bounce a message in a way that is not clear to SES if it is actually permanent, then we don't delete the account.

We could implement our own heuristics for understanding the `status` code and the content of the `diagnosticCode` field, but that may be going down a rabbit hole.

Close this issue as WFM?

**Comment by jrgm** [comment]
(Sub note: the `@mozilla.com` don't log a diagnosticCode, just the `5.0.0`. If I had the `Message-ID`, I could file a ticket with AWS and ask for clarification on why this specific case is not classified as Permanent.)

**Comment by jrgm** [comment]
Of course, we could still do what @vladikoff suggests in https://github.com/mozilla/fxa-content-server/issues/5629#issuecomment-363253121

**Comment by rfk** [comment]
> Of course, we could still do what @vladikoff suggests in #5629 (comment)

From the meeting today, yes, let's go ahead and do this thing: delete unverified accounts on transient of the signup email.

**Comment by rfk** [comment]
I tried to dig into the metrics for this a little, based on our meeting earlier today.  Unfortunately we don't appear to have a flow event corresponding to the "your email bounced, typo?" error message that is supposed to be triggered here, but I was able to see whether users who get a bounce were retrying (and then whether they were succeeding).

So...

Around 6% of newly-created accounts get a subsequent bounce of the verification email:

  https://sql.telemetry.mozilla.org/queries/51142/source#137206

Of those signup flows that get a bounce, around 70% will retry by submitting the signup form again:

  https://sql.telemetry.mozilla.org/queries/51142/source#137207

Of the users that retry after a bounce, around 25% of them will succeed at verifying their account:

  https://sql.telemetry.mozilla.org/queries/51142/source#137208

While of the users that do *not* retry after a bounce, around 10% will still somehow succeed at verifying their account, I presume by fixing a transient bounce condition and re-sending the verification email:

  https://sql.telemetry.mozilla.org/queries/51142/source#137209

So hypothetically, if we can fix this to ensure that users who get a bounce are always prompted to retry, we might be able to pick up an extra 15% in verification rate for the 30% of bounce-getters who don't currently retry, who represent around 2% of our total daily account creation attempts.

I won't put absolute numbers in the bug, but the potential impact here in absolute terms actually doesn't seem very high.  That's mostly due to the fact that retrying the account creation only seems to lead to success around 25% of time.
